src/tree/gpu_hist/gradient_based_sampler.cuh

```diff
@@ -0,0 +1,83 @@
+/*!
+ * Copyright 2019 by XGBoost Contributors
+ */
+#pragma once
+#include <xgboost/base.h>
+#include <xgboost/data.h>
+#include <xgboost/span.h>
+
+#include "../../common/device_helpers.cuh"
+#include "../../data/ellpack_page.cuh"
+
+namespace xgboost {
+namespace tree {
+
+struct GradientBasedSample {
+  /*!\brief Number of sampled rows. */
+  size_t sample_rows;
+  /*!\brief Sampled rows in ELLPACK format. */
+  EllpackPageImpl* page;
+  /*!\brief Gradient pairs for the sampled rows. */
+  common::Span<GradientPair> gpair;
+};
+
+/*! \brief Draw a sample of rows from a DMatrix.
+ *
+ * \see Ke, G., Meng, Q., Finley, T., Wang, T., Chen, W., Ma, W., ... & Liu, T. Y. (2017).
+ * Lightgbm: A highly efficient gradient boosting decision tree. In Advances in Neural Information
+ * Processing Systems (pp. 3146-3154).
+ * \see Zhu, R. (2016). Gradient-based sampling: An adaptive importance sampling for least-squares.
+ * In Advances in Neural Information Processing Systems (pp. 406-414).
+ * \see Ohlsson, E. (1998). Sequential poisson sampling. Journal of official Statistics, 14(2), 149.
+ */
+class GradientBasedSampler {
+ public:
+  enum SamplingMethod {
+    /*! \brief When all rows can fit in GPU memory, no sampling is needed. */
+    kNoSampling,
+    /*! \brief Fixed-sized random sampling, weighted by the absolute value of the gradient. */
+    kSequentialPoissonSampling,
+    /*! \brief This is for comparison purposes only, not recommended for real use. */
+    kUniformSampling
+  };
+
+  explicit GradientBasedSampler(BatchParam batch_param,
+                                EllpackInfo info,
+                                size_t n_rows,
+                                float subsample = 1.0f,
+                                SamplingMethod sampling_method = kDefaultSamplingMethod);
+
+  /*! \brief Sample from a DMatrix based on the given gradients. */
+  GradientBasedSample Sample(common::Span<GradientPair> gpair, DMatrix* dmat);
+
+ private:
+  static const SamplingMethod kDefaultSamplingMethod = kSequentialPoissonSampling;
+
+  GradientBasedSample NoSampling(common::Span<GradientPair> gpair, DMatrix* dmat);
+  GradientBasedSample SequentialPoissonSampling(common::Span<GradientPair> gpair, DMatrix* dmat);
+  GradientBasedSample UniformSampling(common::Span<GradientPair> gpair, DMatrix* dmat);
```

Comment:
is this used someplace, or simply kept so that it can be used for comparison when needed?

Reply:
Not used by real code.